### PR TITLE
[3006.x] Increase coverage for textfsm module

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Increase coverage for textfsm module

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by twangboy — DISMISSED** [review]

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+except ImportError:
+    HAS_LIBS = False
+# pylint: enable=import-error,no-name-in-module,unused-import
+
+
+pytestmark = [
+    pytest.mark.skipif(
+        not HAS_LIBS, reason="Install textfsm to be able to run this test."
+    )
+]
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:24]:
```suggestion
textfsm = pytest.importorskip("textfsm", reason="Install textfsm to be able to run this test.")
```

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 2]
```diff
@@ -0,0 +1,713 @@
...
+                ):
+                    mock_read_data = {
+                        "/index": [mock_open_index],
+                        "/juniper_version_template": [
+                            juniper_version_template_one,
+                            juniper_version_template_two,
+                        ],
+                    }
+                    with patch.object(
+                        clitable, "open", mock_open(read_data=mock_read_data)
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:696]:
```suggestion
                        textfsm.clitable, "open", mock_open(read_data=mock_read_data)
```

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:591]:
```suggestion
                        textfsm.clitable, "open", mock_open(read_data=mock_read_data)
```

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+            with patch.dict(
+                textfsm_mod.__grains__,
+                {"textfsm_platform_grain": "textfsm_platform_grain"},
+            ):
+                with patch.dict(
+                    textfsm_mod.__salt__,
+                    {"cp.cache_dir": MagicMock(return_value="/path/to/cache/")},
+                ):
+                    with patch.object(
+                        clitable, "open", mock_open(read_data=mock_open_index)
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:498]:
```suggestion
                        textfsm.clitable, "open", mock_open(read_data=mock_open_index)
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+
+import salt.modules.textfsm_mod as textfsm_mod
+from tests.support.mock import MagicMock, mock_open, patch
+
+textfsm = pytest.importorskip(
+    "textfsm", reason="Install textfsm to be able to run this test."
+)
+
+
+@pytest.fixture()
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:16]:
```suggestion
@pytest.fixture
```

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+def configure_loader_modules():
+    return {textfsm_mod: {"__opts__": {}}}
+
+
+def test_virttual_virtual():
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:21]:
```suggestion
def test_dunder_virtual():
```

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+                                command="sh ver",
+                                platform="",
+                                output_file="salt://textfsm/juniper_version_example",
+                                textfsm_path="salt://textfsm/",
+                            )
+                            assert ret == {
+                                "out": None,
+                                "result": False,
+                                "comment": "Unable to read from salt://textfsm/juniper_version_example. Please specify a valid file or text.",
+                            }
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:465]:
```suggestion
    with patch.object(textfsm_mod, "HAS_CLITABLE", True), patch.dict(
        textfsm_mod.__opts__, {"textfsm_platform_grain": "textfsm_platform_grain"}
    ), patch.dict(
        textfsm_mod.__grains__,
        {"textfsm_platform_grain": "textfsm_platform_grain"},
    ), patch.dict(
        textfsm_mod.__salt__,
        {"cp.cache_dir": MagicMock(return_value="/path/to/cache/")},
    ), patch.object(
        textfsm_mod.clitable,
        "open",
        mock_open(read_data=mock_open_index),
    ), patch.dict(
        textfsm_mod.__salt__,
        {"cp.get_file_str": MagicMock(return_value=False)},
    ):
        ret = textfsm_mod.index(
            command="sh ver",
            platform="",
            output_file="salt://textfsm/juniper_version_example",
            textfsm_path="salt://textfsm/",
        )
        assert ret == {
            "out": None,
            "result": False,
            "comment": "Unable to read from salt://textfsm/juniper_version_example. Please specify a valid file or text.",
        }
```

Comment (s0undt3ch) [reply tests/pytests/unit/modules/test_textfsm_mod.py:465]: We really don't need to deep nest this...

File: tests/pytests/unit/modules/test_textfsm_mod.py [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+                            ret = textfsm_mod.index(
+                                command="sh ver",
+                                platform="",
+                                textfsm_path="salt://textfsm/",
+                            )
+                            assert ret == {
+                                "out": None,
+                                "result": False,
+                                "comment": "Please specify a valid output text or file",
+                            }
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_textfsm_mod.py:504]: Same as above, too much nesting.

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Basically, only deep nest if we really need to.

**Review by s0undt3ch — DISMISSED** [review]

**Review by twangboy — DISMISSED** [review]

**Review by twangboy — DISMISSED** [review]

**Review by twangboy — APPROVED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Pull request merged**